src/algebra/ring_quot.lean

```diff
@@ -0,0 +1,274 @@
+  one           := quot.mk _ 1,
+  one_mul       := by { rintros ⟨⟩, change quot.mk _ _ = _, rw one_mul, },
+  mul_one       := by { rintros ⟨⟩, change quot.mk _ _ = _, rw mul_one, },
+  left_distrib  := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, change quot.mk _ _ = _, rw left_distrib, refl, },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, change quot.mk _ _ = _, rw right_distrib, refl, }, }
+
+instance {R : Type u₁} [ring R] (r : R → R → Prop) : ring (ring_quot r) :=
+{ neg           := quot.map (λ a, -a)
+    (λ a b h, by simp only [neg_eq_neg_one_mul a, neg_eq_neg_one_mul b, rel.mul_right h]),
+  add_left_neg  := by { rintros ⟨⟩, change quot.mk _ _ = _, rw add_left_neg, refl, },
+  .. (ring_quot.semiring r) }
+
+instance {R : Type u₁} [comm_semiring R] (r : R → R → Prop) : comm_semiring (ring_quot r) :=
+{ mul_comm := by { rintros ⟨⟩ ⟨⟩, change quot.mk _ _ = _, rw mul_comm, refl, }
+  .. (ring_quot.semiring r) }
+
+instance {R : Type u₁} [comm_ring R] (r : R → R → Prop) : comm_ring (ring_quot r) :=
+{ .. (ring_quot.comm_semiring r),
+  .. (ring_quot.ring r) }
+
+instance (s : A → A → Prop) : algebra S (ring_quot s) :=
+{ smul      := λ r a, (quot.mk _ (r • 1)) * a,
+  to_fun    := λ r, quot.mk _ (r • 1),
+  map_one'  := by { rw one_smul, refl, },
+  map_mul'  := λ r s, by { change _ = quot.mk _ _, simp [mul_smul], },
+  map_zero' := by { rw zero_smul, refl, },
+  map_add'  := λ r s, by { change _ = quot.mk _ _, rw add_smul, },
+  commutes' := λ r, by { rintros ⟨a⟩, change quot.mk _ _ = quot.mk _ _, simp, },
+  smul_def' := λ r u, rfl, }
+
+instance (r : R → R → Prop) : inhabited (ring_quot r) := ⟨0⟩
+
+/--
+The quotient map from a ring to its quotient, as a homomorphism of rings.
+-/
+def mk_ring_hom (r : R → R → Prop) : R →+* ring_quot r :=
+{ to_fun := quot.mk _,
+  map_one'  := rfl,
+  map_mul'  := λ a b, rfl,
+  map_zero' := rfl,
+  map_add'  := λ a b, rfl, }
+
+lemma mk_ring_hom_rel {r : R → R → Prop} {x y : R} (w : r x y) :
+  mk_ring_hom r x = mk_ring_hom r y :=
+quot.sound (rel.of w)
+
+lemma mk_ring_hom_surjective (r : R → R → Prop) : function.surjective (mk_ring_hom r) :=
+by { dsimp [mk_ring_hom], rintro ⟨⟩, simp, }
+
+@[ext]
+lemma ring_quot_ext {T : Type u₄} [semiring T] {r : R → R → Prop} (f g : ring_quot r →+* T)
+  (w : f.comp (mk_ring_hom r) = g.comp (mk_ring_hom r)) : f = g :=
+begin
+  ext,
+  rcases mk_ring_hom_surjective r x with ⟨x, rfl⟩,
+  exact (congr_arg (λ h : R →+* T, h x) w), -- TODO should we have `ring_hom.congr` for this?
+end
+
+variables  {T : Type u₄} [semiring T]
+
+/--
+Any ring homomorphism `f : R →+* T` which respects a relation `r : R → R → Prop`
+factors through a morphism `ring_quot r →+* T`.
+-/
+def lift (f : R →+* T) {r : R → R → Prop} (w : ∀ {x y}, r x y → f x = f y) :
+  ring_quot r →+* T :=
+{ to_fun := quot.lift f
+  begin
+    rintros _ _ r,
+    induction r with a b r a b c r r' a b c r r' a b c r r',
+    { exact w r, },
+    { simp [r'], },
+    { simp [r'], },
+    { simp [r'], },
```

Comment:
This `with` seems like a bit of a monster. Perhaps clearer as follows?
```suggestion
    induction r,
    case of : _ _ r { exact w r, },
    case add : _ _ _ _ r' { simp [r'], },
    case mul_left  : _ _ _ _ r' { simp [r'], },
    case mul_right  : _ _ _ _ r' { simp [r'], },
```

Reply:
Nice! I didn't know you could use `case` to rename induction hypotheses. Thanks.